src/main/drivers/compass/compass_ist8310.c

```diff
@@ -0,0 +1,256 @@
+ *      0: disable
+ *      1: enable
+ * 2    DRP: DRDY pin polarity control
+ *      0: active low
+ *      1: active high
+ * 1    0   Reserved.
+ * 0    SRST: Soft reset, perform Power On Reset (POR) routine
+ *      0: no action
+ *      1: start immediately POR routine
+ *      This bit will be set to zero after POR routine
+ */
+
+#define IST8310_REG_DATA 0x03
+#define IST8310_REG_WHOAMI 0x00
+
+#define IST8310_REG_STAT1 0x02
+#define IST8310_REG_STAT2 0x09
+
+#define IST8310_DRDY_MASK 0x01
+
+// I2C Control Register
+#define IST8310_REG_CNTRL1 0x0A
+#define IST8310_REG_CNTRL2 0x0B
+#define IST8310_REG_AVERAGE 0x41
+#define IST8310_REG_PDCNTL 0x42
+
+// Parameter
+// ODR = Output Data Rate, we use single measure mode for getting more data.
+#define IST8310_ODR_SINGLE 0x01
+#define IST8310_ODR_10_HZ 0x03
+#define IST8310_ODR_20_HZ 0x05
+#define IST8310_ODR_50_HZ 0x07
+#define IST8310_ODR_100_HZ 0x06
+
+// Device ID (ist8310 -> 0x10)
+#define IST8310_CHIP_ID 0x10
+#define IST8310_AVG_16  0x24
+#define IST8310_PULSE_DURATION_NORMAL 0xC0
+
+#define IST8310_CNTRL2_RESET 0x01
+#define IST8310_CNTRL2_DRPOL 0x04
+#define IST8310_CNTRL2_DRENA 0x08
+
+static bool ist8310Init(magDev_t *magDev)
+{
+    extDevice_t *dev = &magDev->dev;
+    uint8_t regTemp;
+
+    busDeviceRegister(dev);
+
+    // soft-Reset
+    bool ack = busReadRegisterBuffer(dev, IST8310_REG_CNTRL2, &regTemp, 1);
+    regTemp |= IST8310_CNTRL2_RESET;
+    ack = ack && busWriteRegister(dev, IST8310_REG_CNTRL2, regTemp);
+    delay(30);
+
+    // Init setting : avg16 / pulse mode
+    ack = ack && busWriteRegister(dev, IST8310_REG_AVERAGE, IST8310_AVG_16);
+    delay(5);
+    ack = ack && busWriteRegister(dev, IST8310_REG_PDCNTL, IST8310_PULSE_DURATION_NORMAL);
+    delay(5);
+
+    // DRDY enable
+    ack = ack && busReadRegisterBuffer(dev, IST8310_REG_CNTRL2, &regTemp, 1);
+    regTemp |= IST8310_CNTRL2_DRENA;
+    ack = ack && busWriteRegister(dev, IST8310_REG_CNTRL2, regTemp);
+    delay(5);
+
+    // DRDY polarity
+    ack = ack && busReadRegisterBuffer(dev, IST8310_REG_CNTRL2, &regTemp, 1);
+    regTemp &= ~IST8310_CNTRL2_DRPOL;
+    ack = ack && busWriteRegister(dev, IST8310_REG_CNTRL2, regTemp);
+    delay(5);
+
+    return ack;
+}
+
+static bool ist8310Read(magDev_t * magDev, int16_t *magData)
+{
+    extDevice_t *dev = &magDev->dev;
+
+    uint8_t buf[6];
+    // uint8_t LSB2FSV = 3; // 3mG - 14 bit
+    static uint8_t status;
+    static enum {
+        STATE_INIT_DATA,
+        STATE_REQUEST_DATA,
+        STATE_READ_STATUS,
+        STATE_WAIT_STATUS,
+        STATE_WAIT_READ,
+    } state = STATE_INIT_DATA;
+
+    switch (state) {
+        default:
+        case STATE_INIT_DATA:
+            magData[X] = 0;
+            magData[Y] = 0;
+            magData[Z] = 0;
+
+            state = STATE_REQUEST_DATA;
+
+            return false;
+        case STATE_REQUEST_DATA:
+            busWriteRegister(dev, IST8310_REG_CNTRL1, IST8310_ODR_SINGLE);
+            delay(6);
```

Comment:
I think is better to remove this delay. I understand the reason of it, it's a wait to let the mag get 16 samples, that is what we are averaging.
But this wait makes the code "block", wasting time. The states are just to let the code do other things between states. I think remember the MAG task is called at 5Hz, so it will have a wait of 200ms between states, more than enough, automatically.

Reply:
Looking again at this, maybe we have added too much states. 5 states at 5Hz will be 1 update at second only if I'm not wrong, right? Maybe we need only two states, one asking for the data, and the other reading and processing it.
I will test how it behaves later, to confirm that.

Reply:
Oh, it seems you are using `busReadRegisterBufferStart` and not `busReadRegisterBuffer`. I suppose the first one does not "block" until the result is there.